walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,96 @@
-  }
+
+    /**
+    * Returns a collection of available TimeRange given a collection of Event objects and
+    * MeetingRequest. A TimeRange is available if the person in MeetingRequest are all available.
+    * We first do this by checking which TimeRange are the attendees not available in and 
+    * use sliding window to record the available TimeRange provided they satisfy the duration.
+    * @param    events  a collection of Event objects
+    * @param    request a MeetingRequest object
+    * @return           collection of available TimeRange  
+    */
+    public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+        ArrayList<TimeRange> overlappingTime = new ArrayList<TimeRange>();
+        ArrayList<TimeRange> mergedEventTime = new ArrayList<TimeRange>();
+        ArrayList<TimeRange> availableTime = new ArrayList<TimeRange>();
+        Set<String> requestAttendees = new HashSet<String>(request.getAttendees());
+
+        // record TimeRange whereby any of attendees is unavailable, i.e. in another event
+        for(Event event: events){
+            Set<String> eventAttendees = new HashSet<String>(event.getAttendees());
+            eventAttendees.retainAll(requestAttendees);
+            if (!eventAttendees.isEmpty()) {
+                overlappingTime.add(event.getWhen());
+            }
+        }
+
+        // merge eventTime time range
+        mergedEventTime = mergeEventTime(overlappingTime);
+
+        // sliding window to find available timing that matches the duration
+        int prevEnd = TimeRange.START_OF_DAY;
+        long duration=request.getDuration();
+
+        // check from START_OF_DAY and invertals within each event
+        for(TimeRange curEvent: mergedEventTime){
+            if (curEvent.start() - prevEnd >= duration){
+                availableTime.add(TimeRange.fromStartEnd(prevEnd, curEvent.start(), false));
+            }
+            prevEnd = curEvent.end();
+        }
+
+        // check from last event to END_OF_DAY
+        if (TimeRange.END_OF_DAY - prevEnd >= duration) {
+            availableTime.add(TimeRange.fromStartEnd(prevEnd, TimeRange.END_OF_DAY, true));
+        }
+
+        return availableTime;
+    }
+
+    /**
+    * This function will merge overlapping intervals. We do this by first sorting event time by 
+    * their start time, and try to merge greedily as long as the next event time is still 
+    * overlapping. 
+    * @param    events  a collection of Event objects
+    * @return           collection of merged event time, sorted by start
+    */
+    private ArrayList<TimeRange> mergeEventTime(ArrayList<TimeRange> eventTime){
+        if (eventTime.size() < 1) {return eventTime;}
+
+        Collections.sort(eventTime, TimeRange.ORDER_BY_START);
+        ArrayList<TimeRange> mergedEventTime = new ArrayList();  
+        int prevStart = eventTime.get(0).start(), prevEnd = eventTime.get(0).end();
```

Comment:
Good naming of variable names here.

A small suggestion in style, separating the two variable initialisation into two lines.
Usually the only situation we initialise multiple variables in one single line is when they are super tightly coupled, and initialised only using simple values. For example, (x, y, z) when they represent the 3 component of a coordinate. Benefits in doing so:
* you can put breakpoints in each line.
* It is easier to visually read, compare, and tell the difference in initialising each variable.

Reply:
ah i see. usually i make them a one-liner if they have the same type to reduce line count. but i guess breaking them will improve readability if the initialization is too complex. thanks!

Reply:
Thanks. Just a suggestion, reducing line count is usually the least important factor to balance, in terms of weight when considering different factors, I recommend this order: the ability to place breakpoints > readability >>> line count. 

For example sometimes it is perfectly fine to put 3 blank lines just to make your code structurally "looks" like your logical structure.